File: frontend/src/types/portfolio.ts
/** Portfolio types for FR-PF001 ~ PF003. */

export type BoardType = "main" | "chinext" | "star"

export interface Position {
  id: string
  symbol: string
  name: string
  board: BoardType
  costPrice: number
  shares: number
  buyDate: string
  note?: string
}

export interface Portfolio {
  version: 1
  updatedAt: string
  positions: Position[]
}

export interface PositionWithPnL extends Position {
  currentPrice: number | null
  todayPctChange: number | null
  marketValue: number
  pnl: number
  pnlPercent: number
}

export interface PortfolioSummary {
  positionCount: number
  totalCost: number
  totalMarketValue: number
  totalPnL: number
  totalPnLPercent: number
  positions: PositionWithPnL[]
}

export interface PositionAdvice {
  symbol: string
  name: string
  action: string
  reason: string
  target_price: number | null
}

export interface ConcentrationRisk {
  level: "low" | "medium" | "high"
  description: string
  top_holdings_pct: number | null
}

export interface PortfolioDiagnosis {
  status: string
  health_score: number
  health_label: string
  summary: string
  concentration_risk: ConcentrationRisk | null
  position_advice: PositionAdvice[]
  rebalancing: string[]
  risk_warnings: string[]
  reasoning: string[]
  generated_at: string | null
  model_used: string | null
  message?: string
}

File: src/web/schemas/portfolio.py
"""Portfolio Pydantic models."""

from __future__ import annotations

from pydantic import BaseModel, Field


class PortfolioPosition(BaseModel):
    """A portfolio position for persistence (matches frontend Position type)."""

    id: str
    symbol: str
    name: str
    board: str = "main"
    costPrice: float
    shares: int
    buyDate: str = ""
    note: str | None = None


class PortfolioData(BaseModel):
    """Full portfolio payload for save/load."""

    version: int = 1
    updatedAt: str = ""
    positions: list[PortfolioPosition] = Field(default_factory=list)


class PortfolioPositionInput(BaseModel):
    """A single portfolio position submitted by the frontend."""

    symbol: str
    name: str
    board: str = "main"
    cost_price: float
    shares: int
    buy_date: str | None = None
    current_price: float | None = None
    pnl: float | None = None
    pnl_percent: float | None = None


class PortfolioDiagnosisRequest(BaseModel):
    """Request body for portfolio AI diagnosis."""

    positions: list[PortfolioPositionInput]


class PositionAdviceItem(BaseModel):
    """AI advice for a single position."""

    symbol: str
    name: str
    action: str = "hold"
    reason: str = ""
    target_price: float | None = None


class ConcentrationRisk(BaseModel):
    """Portfolio concentration risk assessment."""

    level: str = "low"
    description: str = ""
    top_holdings_pct: float | None = None


class PortfolioDiagnosisResult(BaseModel):
    """Full AI portfolio diagnosis result."""

    status: str = "success"
    health_score: int = 50
    health_label: str = "一般"
    summary: str = ""
    concentration_risk: ConcentrationRisk | None = None
    position_advice: list[PositionAdviceItem] = Field(default_factory=list)
    rebalancing: list[str] = Field(default_factory=list)
    risk_warnings: list[str] = Field(default_factory=list)
    reasoning: list[str] = Field(default_factory=list)
    generated_at: str | None = None
    model_used: str | None = None
    message: str | None = None
